pkg/webhook/admission/deployment/model_provisioning_injector.go

```diff
@@ -0,0 +1,171 @@
+/*
+Copyright 2019 kubeflow.org.
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+	http://www.apache.org/licenses/LICENSE-2.0
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package deployment
+
+import (
+	"fmt"
+	"strings"
+
+	"github.com/kubeflow/kfserving/pkg/constants"
+
+	appsv1 "k8s.io/api/apps/v1"
+	v1 "k8s.io/api/core/v1"
+)
+
+const (
+	ModelProvisioningContainerName    = "model-provisioner"
+	ModelProvisioningVolumeName       = "kfserving-provision-location"
+	ModelProvisioningContainerImage   = "kcorer/kfdownloader"
+	ModelProvisioningContainerVersion = "latest"
+	PvcURIPrefix                      = "pvc://"
+	PvcSourceMountName                = "kfserving-pvc-source"
+	PvcSourceMountPath                = "/mnt/pvc"
+)
+
+// InjectModelProvisioner injects an init container to provision model data
+// for the serving container in a unified way across storage tech by injecting
+// a provisioning INIT container. This is a work around because KNative does not
+// support INIT containers: https://github.com/knative/serving/issues/4307
+func InjectModelProvisioner(deployment *appsv1.Deployment) error {
+
+	var srcURI, mountPath string
+
+	annotations := deployment.Spec.Template.ObjectMeta.Annotations
+	podSpec := &deployment.Spec.Template.Spec
+
+	// Only inject if the required annotations are set
+	if _, ok := annotations[constants.KFServiceModelProvisioningSourceURIAnnotationKey]; ok {
+		srcURI = annotations[constants.KFServiceModelProvisioningSourceURIAnnotationKey]
+	} else {
+		return nil
+	}
+
+	if _, ok := annotations[constants.KFServiceModelProvisioningMountPathAnnotationKey]; ok {
+		mountPath = annotations[constants.KFServiceModelProvisioningMountPathAnnotationKey]
+	} else {
+		mountPath = constants.DefaultModelLocalMountPath
+	}
+
+	// Only inject provisioning for supported URIs
```

Comment:
I thought we were going to run the injector in all cases? If it isn't a supported URI, our validating admission controller should reject it. Otherwise we're going to need to keep editing this code every time we add new storage support.

Reply:
Moved validation to the validator